### ISSUE TITLE
SHGetFileInfo throws "Pointers cannot reference marshaled structures. Use ByRef instead."

### ISSUE THREAD

**Issue by AliveDevil** [microsoft/CsWin32#361]
## Actual behavior

Parameter #&#8203;3 cannot be marshalled: Pointer cannot reference marshaled structures.
Attached is a reproducing example: [repro.zip](https://github.com/microsoft/CsWin32/files/7000486/repro.zip)

## Repro steps

1. `NativeMethods.txt` content:
```
SHGetFileInfo
```

2. `NativeMethods.json` content (if present):
```json
{
    "$schema": "https://aka.ms/CsWin32.schema.json",
    "public": true,
    "emitSingleFile": true
}

```

3. Any of your own code that should be shared?
```c#
using System.Runtime.CompilerServices;
using Storage.FileSystem;
using UI.Shell;

partial class PInvoke
{
    public static unsafe nuint SHGetFileInfo(string pszPath, FILE_FLAGS_AND_ATTRIBUTES dwFileAttributes, in SHFILEINFOW sfi, SHGFI_FLAGS uFlags)
    {
        var psfi = (SHFILEINFOW*)Unsafe.AsPointer(ref Unsafe.AsRef(sfi));
        fixed (char* pszPathLocal = pszPath)
        {
            return SHGetFileInfo(pszPathLocal, dwFileAttributes, psfi, (uint)Unsafe.SizeOf<SHFILEINFOW>(), uFlags);
        }
    }
}
```

### Context

- CsWin32 version: `0.1.506-beta`
- Target Framework: `net47`

**Comment by AArnott** [comment]
I am not seeing this exception: Here's my console app:

```cs
unsafe
{
    SHFILEINFOW foo = default;
    nuint h = SHGetFileInfo(@"c:\temp", default, foo, default);
    Console.WriteLine("hi");
}

namespace Windows.Win32
{
    partial class PInvoke
    {
        internal static unsafe nuint SHGetFileInfo(string pszPath, FILE_FLAGS_AND_ATTRIBUTES dwFileAttributes, in SHFILEINFOW sfi, SHGFI_FLAGS uFlags)
        {
            var psfi = (SHFILEINFOW*)Unsafe.AsPointer(ref Unsafe.AsRef(sfi));
            fixed (char* pszPathLocal = pszPath)
            {
                return SHGetFileInfo(pszPathLocal, dwFileAttributes, psfi, (uint)Unsafe.SizeOf<SHFILEINFOW>(), uFlags);
            }
        }
    }
}
```

Does this work for you?

**Comment by AArnott** [comment]
Ah, I was running on `net6.0`. But you mentioned it fails on `net47`. I can repro on `net472`.

**Comment by AArnott** [comment]
The `char` field in the struct is causing it to require marshaling. We'll either need to find a way to avoid that, or fix CsWin32 to recognize that `char` has this impact so that APIs that use it can use `ref` instead of pointers.

**Comment by AliveDevil** [comment]
Just tested on .NET 4.8, fails there as well - so general .NET Framework Runtime issue I guess.

Before in `0.1.422` it worked as the `char`-field was a `ushort`-array. Is there any chance that there could be a custom mapping for .NET Framework to map such fixed char arrays to fixed ushorts?
Or opt-in to the `char`-mapping for .NET Standard 2.1+ and any .NET Core/.NET 5+-version.

**Comment by AArnott** [comment]
That's a keen observation. If nothing else, we could emit code where we use `ushort` for the (private) field but expose a public property typed as `char` that reads/writes the field.

**Comment by AliveDevil** [comment]
As the documentation mentions that chars are sometimes blittable for .NET Standard (I read: .NET Core and beyond) (https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices) I'd go the route of least support and make it a .NET Framework compatible blittable type (https://docs.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types)

By any chance: could there be an option for CSWin32 to ignore certain types in code gen (like this struct) so I as a developer could add my own types which CsWin32 would just use as-is?
So the pinvoke calls are generated but the underlying struct is provided by the implementing project to workaround these runtime issues?

**Comment by AArnott** [comment]
From the doc you linked to:

> char is blittable in a one-dimensional array or if it's part of a type that contains it's explicitly marked with [StructLayout] with CharSet = CharSet.Unicode.
`[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]`

I just tested it, and .NET Framework is fixed when this attribute is applied to the structs that contain the `char` fields. That's super-easy to do, so I'll make the fix to the source generator that way.

> Just saw b1be7ba. Is that already available in 0.1.506?

Yes, it's been around for quite a while.

**Comment by AliveDevil** [comment]
I see. I can confirm adding the StructLayout-attribute to the the nested structs in `SHFILEINFOW` does fix this down to .NET 4.6 (didn't test below that).

Interesting though that the .NET Framework docs mention that char is not blittable, but the standard docs extend that to StructLayout-annotated structs with CharSet specification.

**Comment by AArnott** [comment]
This was fixed in db245d9d5b3c3be8cde296a0350be4e8ada91e5f as part of #389.

**Comment by E021ntox** [comment]
@AArnott 

SHGetFileInfo lacks this signature.

https://www.pinvoke.net/default.aspx/shell32.shgetfileinfo
```
[StructLayout(LayoutKind.Sequential, CharSet=CharSet.Auto)]
public struct SHFILEINFO
{
     **public IntPtr hIcon;**
     public int iIcon;
     public uint dwAttributes;
     [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 260)]
     public string szDisplayName;
     [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 80)]
     public string szTypeName;
};

```

https://github.com/microsoft/win32metadata/blob/ad21d248aeaad1418f1a2e9e266b56c559ecd932/generation/WinSDK/RecompiledIdlHeaders/um/shellapi.h#L1213
